### PR TITLE
Rewrite gzip component to support trailing bytes without external tool and compress with PIGZ. Addresses #476

### PR DESCRIPTION
**Pull request by alchzh**
- [x] I have reviewed the [OFRAK contributor guide](https://ofrak.com/docs/contributor-guide/getting-started.html) and attest that this pull request is in accordance with it.

Rewrite gzip component to support trailing bytes without external tool. 

**Link to Related Issue(s)**

#476
python/cpython#68489

**Please describe the changes in your request.**

- Replaces the use of the `gzip` library with calls to the underlying `zlib` library directly, in a way that doesn't fail with trailing garbage at the end of files.
- Substantially rewrites of gzip test cases to use a common template
  - adds a test for multi member gzip files
  - the trailing bytes test actually tests the unpack

### CODE REVIEW

File: ofrak_core/test_ofrak/components/test_gzip_component.py
Comment (rbs-jacob) [inline ofrak_core/test_ofrak/components/test_gzip_component.py:16]:
```suggestion
    """
    Template for tests that test different inputs the gzip component should support
    unpacking.
    """
```

Minor, but FYI for future that we format docstrings in a specific way. [More info in the OFRAK contributor style guide docs.](https://ofrak.com/docs/contributor-guide/getting-started.html#functions-and-methods)

File: ofrak_core/ofrak/core/gzip.py
Comment (rbs-jacob) [inline ofrak_core/ofrak/core/gzip.py:78]: Probably makes sense to break each case of this `if` into its own helper function. 

The reason is that we mandate 100% function coverage in testing before code can be merged, but we have no requirements about line coverage or branch coverage. We want to make sure that we test both of these cases, with no possibility of regressions in the future. As such, if each case is its own function, our function coverage testing will confirm that both branches are hit. 

I suspect that all of our test files are small, and that the `pigz` case is never tested.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]

File: ofrak_core/ofrak/core/gzip.py
Comment (alchzh) [reply ofrak_core/ofrak/core/gzip.py]: Refactored into separate functions and added test case with a big binary for coverage

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
This needs a changelog entry, but after that, should probably be good to go.

File: ofrak_core/test_ofrak/components/test_gzip_component.py
Comment (whyitfor) [inline ofrak_core/test_ofrak/components/test_gzip_component.py:50]: This should be `raise NotImplementedError()`: https://ofrak.com/docs/contributor-guide/getting-started.html#raising-exceptions.

File: ofrak_core/test_ofrak/components/test_gzip_component.py
Comment (whyitfor) [inline ofrak_core/test_ofrak/components/test_gzip_component.py:89]: Could you just concatenate DEADBEEF on this line?

File: ofrak_core/test_ofrak/components/test_gzip_component.py
Comment (whyitfor) [reply ofrak_core/test_ofrak/components/test_gzip_component.py:50]: Should it also be marked as an `@abstractmethod`?

File: ofrak_core/ofrak/core/gzip.py
Comment (whyitfor) [inline ofrak_core/ofrak/core/gzip.py:51]: Since `unpack_with_zlib_module` now handles the trailing bytes, I'm not following why we are catching exceptions here?

It seems like this logic could be simplified to:

```suggestion
        pigz_installed = await PIGZ.is_tool_installed()
        if pigz_installed and len(data) >= 1024 * 1024 * 4:
            uncompressed_data = await self.unpack_with_pigz(data)
        else:
            uncompressed_data = await self.unpack_with_zlib_module(data)
```

File: ofrak_core/ofrak/core/gzip.py
Comment (whyitfor) [inline ofrak_core/ofrak/core/gzip.py:98]: Why is this change necessary? What was wrong with the Packer as it was implemented?

**Review by whyitfor — CHANGES_REQUESTED** [review]
See comments.

File: ofrak_core/ofrak/model/component_model.py
Comment (whyitfor) [inline ofrak_core/ofrak/model/component_model.py:21]: I'd rather not change this.

The `ComponentExternalTool` is a nice, simple, stateless class. Can you cache the tool installation state in the `GzipUnpacker`? Until we have more use cases for caching this result I'd rather keep this class simple and immutable.

File: ofrak_core/test_ofrak/components/test_gzip_component.py
Comment (alchzh) [reply ofrak_core/test_ofrak/components/test_gzip_component.py:89]: no, that would add it to the compressed data section instead of writing extra trailing bytes.

File: ofrak_core/ofrak/core/gzip.py
Comment (alchzh) [reply ofrak_core/ofrak/core/gzip.py:51]: I had that originally, but I figured in case of some other format incompatibility or big in python we're not aware of might as well have the fallback still.

File: ofrak_core/ofrak/core/gzip.py
Comment (alchzh) [reply ofrak_core/ofrak/core/gzip.py:98]: using `zlib` is much faster, avoids a bunch of unnecessary processing in GzipFile

test on Silesia corpus:
https://gist.github.com/alchzh/a46140e683ff5e05be8d6b027a4f3fba

almost 4x faster on "mozilla" which is a 50MB tarfile of a Linux application so probably most applicable for us (4.4s for GzipFile vs 1.2s for zlib)

Comment (alchzh) [reply ofrak_core/ofrak/core/gzip.py:98]: on that note, we should probably be using pigz for packing large files as well

File: ofrak_core/ofrak/core/gzip.py
Comment (alchzh) [reply ofrak_core/ofrak/core/gzip.py]: no longer using PIGZ for decompression

**Comment by alchzh** [comment]
Based on benchmarks of zlib module vs pigz, no benefit to using pigz for decompression, but significant benefit to using it for compression.

 https://gist.github.com/alchzh/24ca62a6f87548ea2b1911c174e0ed54

File: ofrak_core/ofrak/core/gzip.py
Comment (whyitfor) [inline ofrak_core/ofrak/core/gzip.py:90]: So basically this will run even if PIGZ isn't installed; OFRAK does not "track" this third party dependency.

This seems like an OK tradeoff; we can put it in our Docker image, and people can use this even if it is not installed.

(Writing this mainly to document this for posterity).

**Review by whyitfor — APPROVED** [review]
Looks good to me.

Especially like the profiling and data-driven approach.

File: ofrak_core/CHANGELOG.md
Comment (whyitfor) [inline ofrak_core/CHANGELOG.md:34]:
```suggestion
- In `GzipUnpacker`, use the standard python `zlib` library to compress small files and decompress all files. Use `pigz` if it is installed to compress files 1MB and larger. ([#472](https://github.com/redballoonsecurity/ofrak/pull/472) and [#485](https://github.com/redballoonsecurity/ofrak/pull/485))
```

**Review by rbs-jacob — APPROVED** [review]
I did not review the tests in great detail, but I like the way that this implementation ended up. Seems like a big improvement!

File: ofrak_core/ofrak/core/gzip.py
Comment (rbs-jacob) [inline ofrak_core/ofrak/core/gzip.py:71]: I think this comment is very unclear as it stands. My understanding is that it means "this is why we're using `zlib.decompressobj`," but it reads like "this might not work in Python 3.11 because we aren't using a `zlib.Decompress` object."

I would also amend the comment to cite a source (or include justification) for why we're doing decompression this way here. To a future reader, I don't think it will be obvious why we're decompressing data that begins with these magic bytes in chunks like this.

**Review by rbs-jacob — CHANGES_REQUESTED** [review]
I relented and included one minor change that I would like to see before merging this.

**Review by rbs-jacob — APPROVED** [review]
Thanks for this fix. Looks ready to merge once the tests pass!

**Pull request merged**
